### PR TITLE
Cassandra 16958

### PR DESCRIPTION
**Pull request by jmckenzie-dev**

### CODE REVIEW

File: src/java/org/apache/cassandra/auth/AuthCacheService.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+            cache.warm();
+        }
+    }
+
+    /**
+     * We have a couple of static initializer functions to create caches scattered across various classes, some solo
+     * and some with multiple member variables. As we expect these caches to be created and initialized in one logical
+     * block, we tie them together and use them here.
+     *
+     * Note: We also register the PasswordAuthenticator cache here with an instance member call to:
```
Comment (maedhroz) [inline src/java/org/apache/cassandra/auth/AuthCacheService.java]: nit: Where "here" means "in `StorageService#doAuthSetup()` before calling `initializeAndRegisterCaches()`"?

File: test/unit/org/apache/cassandra/service/ClientStateTest.java [hunk 2]
```diff
@@ -22,18 +22,12 @@
 import java.util.concurrent.atomic.AtomicInteger;
 
 import com.google.common.collect.Iterables;
+import org.apache.cassandra.auth.*;
```
Comment (maedhroz) [inline test/unit/org/apache/cassandra/service/ClientStateTest.java]: nit: Probably want to at least move this back down w/ the other C* imports, if not move and get rid of the wildcard as well.

File: src/java/org/apache/cassandra/auth/AuthCacheService.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+import com.google.common.base.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class AuthCacheService
+{
+    private static final Logger logger = LoggerFactory.getLogger(AuthCacheService.class);
+    public static final AuthCacheService instance = new AuthCacheService();
+
+    private final Set<AuthCache<?, ?>> caches = new HashSet<>();
```
Comment (maedhroz) [inline src/java/org/apache/cassandra/auth/AuthCacheService.java:38]: A few questions around this `caches` state and how it's protected...

The first thing I noticed is that `warmCaches()` isn't synchronized, so we might be left open to a `ConcurrentModificationException`. Then I started looking at the actual sources of possible concurrent access to `caches`. It seems like the primary startup pathway isn't really a concern, but the `nodetool` commands for resuming bootstrap and join might be, since they hit `doAuthSetup()`. It that's the case, the `warmCaches()` issue above stands, but also only `Roles.init()` seems to protect us against double-adding its `cache`. (For `AuthenticatedUser`, that would be `permissionsCache` and `networkPermissionsCache`.) Then again, registering a few more things to `caches` on those `nodetool` commands might no matter, given `warmCaches()` is only called from `CassandraDaemon#setup()`. Finally, `Roles.init()` and `PasswordAuthenticator#setup()` aren't synchronized, so their `cache` fields aren't protected.

Comment (maedhroz) [reply src/java/org/apache/cassandra/auth/AuthCacheService.java:38]: ...and as I finished writing that, I see that `doAuthSetup()` is guarded by `authSetupCalled` and should only happen once :)

So unless we're going to hit `warmCaches()` in `CassandraDaemon#setup()` somehow after a `nodetool` command, I'm not sure if there's a thread-safety issue at all. If that *is* possible, just throwing `Collections.synchronizedSet()` on `AuthCacheService#caches` would probably be sufficient (without synchronizing any methods here).

File: src/java/org/apache/cassandra/config/Config.java [hunk 4]
```diff
@@ -522,6 +522,8 @@
      */
     public volatile int validation_preview_purge_head_start_in_sec = 60 * 60;
 
+    public boolean auth_cache_warming_enabled = true;
```
Comment (maedhroz) [inline src/java/org/apache/cassandra/config/Config.java]: nit: Did the original Jira description mention making this opt-in?

File: test/unit/org/apache/cassandra/tools/nodetool/InvalidateJmxPermissionsCacheTest.java [hunk 5]
```diff
@@ -21,15 +21,11 @@
 import java.util.Set;
 import javax.security.auth.Subject;
 
+import org.apache.cassandra.auth.*;
```
Comment (maedhroz) [inline test/unit/org/apache/cassandra/tools/nodetool/InvalidateJmxPermissionsCacheTest.java]: nit: Move this down to the right place (and consider individual imports).

File: src/java/org/apache/cassandra/service/StorageService.java [hunk 6]
```diff
@@ -60,6 +60,7 @@
 import com.google.common.collect.*;
 import com.google.common.util.concurrent.*;
 
+import org.apache.cassandra.auth.*;
```
Comment (maedhroz) [inline src/java/org/apache/cassandra/service/StorageService.java]: nit: Consider restoring individual imports.

File: test/unit/org/apache/cassandra/tools/nodetool/InvalidateNetworkPermissionsCacheTest.java [hunk 7]
```diff
@@ -18,12 +18,11 @@
 
 package org.apache.cassandra.tools.nodetool;
 
+import org.apache.cassandra.auth.*;
```
Comment (maedhroz) [inline test/unit/org/apache/cassandra/tools/nodetool/InvalidateNetworkPermissionsCacheTest.java]: nit: Move this down to the right place (and consider individual imports).

File: test/unit/org/apache/cassandra/tools/nodetool/InvalidatePermissionsCacheTest.java [hunk 8]
```diff
@@ -24,19 +24,12 @@
 import java.util.List;
 import java.util.Set;
 
+import org.apache.cassandra.auth.*;
```
Comment (maedhroz) [inline test/unit/org/apache/cassandra/tools/nodetool/InvalidatePermissionsCacheTest.java]: nit: Move this down to the right place (and consider individual imports).

File: test/unit/org/apache/cassandra/tools/nodetool/InvalidateNetworkPermissionsCacheTest.java [hunk 9]
```diff
@@ -46,6 +45,9 @@ public static void setup() throws Exception
 
         roleManager.createRole(AuthenticatedUser.SYSTEM_USER, ROLE_A, AuthTestUtils.getLoginRoleOptions());
         roleManager.createRole(AuthenticatedUser.SYSTEM_USER, ROLE_B, AuthTestUtils.getLoginRoleOptions());
+        roleManager.createRole(AuthenticatedUser.SYSTEM_USER, ROLE_A, AuthTestUtils.getLoginRoleOptions());
+        roleManager.createRole(AuthenticatedUser.SYSTEM_USER, ROLE_B, AuthTestUtils.getLoginRoleOptions());
```
Comment (maedhroz) [inline test/unit/org/apache/cassandra/tools/nodetool/InvalidateNetworkPermissionsCacheTest.java]: Do we need to add these roles twice? The test still seems to pass if I remove the duplicates...

File: test/unit/org/apache/cassandra/tools/nodetool/InvalidateRolesCacheTest.java [hunk 10]
```diff
@@ -18,12 +18,14 @@
 
 package org.apache.cassandra.tools.nodetool;
 
+import org.apache.cassandra.auth.AuthCacheService;
```
Comment (maedhroz) [inline test/unit/org/apache/cassandra/tools/nodetool/InvalidateRolesCacheTest.java]: nit: Move down alongside other C* imports.

File: test/unit/org/apache/cassandra/cql3/CQLTester.java [hunk 11]
```diff
@@ -46,6 +46,7 @@
 import com.google.common.base.Strings;
 import com.google.common.collect.ImmutableSet;
 
+import org.apache.cassandra.auth.AuthCacheService;
 import org.apache.cassandra.exceptions.InvalidRequestException;
 import org.apache.cassandra.io.util.File;
```
Comment (maedhroz) [inline test/unit/org/apache/cassandra/cql3/CQLTester.java]: super nit: These three imports should move down with the others of their kind :)

File: test/unit/org/apache/cassandra/auth/AuthCacheTest.java [hunk 12]
```diff
@@ -226,11 +213,80 @@ public void testCassandraExceptionPassThroughWhenCacheRefreshed() throws Interru
         }
     }
 
+@Test
```
Comment (maedhroz) [inline test/unit/org/apache/cassandra/auth/AuthCacheTest.java]: nit: indentation

File: test/unit/org/apache/cassandra/auth/AuthCacheTest.java [hunk 13]
```diff
@@ -226,11 +213,80 @@ public void testCassandraExceptionPassThroughWhenCacheRefreshed() throws Interru
...
+        isCacheEnabled = false;
+        TestCache cache = newCache(() -> Collections.singletonMap("0", 0));
+        cache.warm();
+        assertEquals(0, cache.getEstimatedSize());
+    }
+
+    @Test
+    public void providerSuppliesMoreEntriesThanCapacity()
+    {
+        final int maxEntries = 10;
```
Comment (maedhroz) [inline test/unit/org/apache/cassandra/auth/AuthCacheTest.java]: nit: Minor readability point...had to find this is enforced as a hard-coded 10 in `TestCache` below.

File: test/unit/org/apache/cassandra/auth/CassandraAuthorizerTest.java [hunk 14]
```diff
@@ -35,6 +35,7 @@
     @BeforeClass
     public static void setupClass()
     {
+        System.setProperty("org.apache.cassandra.disable_mbean_registration", "true");
```
Comment (maedhroz) [inline test/unit/org/apache/cassandra/auth/CassandraAuthorizerTest.java]:
```suggestion
               CassandraRelevantProperties.ORG_APACHE_CASSANDRA_DISABLE_MBEAN_REGISTRATION.setBoolean(true);
```

File: test/unit/org/apache/cassandra/auth/CassandraAuthorizerTruncatingTests.java [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+
+/**
+ * For Authorizer based tests where we need to fully truncate the roles, members, and permissions between tests
+ */
+public class CassandraAuthorizerTruncatingTests extends CQLTester
+{
+    @BeforeClass
+    public static void setupClass()
+    {
+        System.setProperty("org.apache.cassandra.disable_mbean_registration", "true");
```
Comment (maedhroz) [inline test/unit/org/apache/cassandra/auth/CassandraAuthorizerTruncatingTests.java]:
```suggestion
        CassandraRelevantProperties.ORG_APACHE_CASSANDRA_DISABLE_MBEAN_REGISTRATION.setBoolean(true);
```

File: test/unit/org/apache/cassandra/auth/CassandraRoleManagerTest.java [hunk 16]
```diff
@@ -85,4 +103,75 @@ private void fetchRolesAndCheckReadCount(IRoleManager roleManager, RoleResource
         long after = getRolesReadCount();
         assertEquals(granted.size(), after - before);
     }
+
+    @Test
+    public void confirmFastRoleSetup() throws Exception
```
Comment (maedhroz) [inline test/unit/org/apache/cassandra/auth/CassandraRoleManagerTest.java]: nit: could remove `throws Exception`

File: test/unit/org/apache/cassandra/auth/CassandraRoleManagerTest.java [hunk 17]
```diff
@@ -85,4 +103,75 @@ private void fetchRolesAndCheckReadCount(IRoleManager roleManager, RoleResource
...
+    {
+        IRoleManager roleManager = new AuthTestUtils.LocalCassandraRoleManager();
+        roleManager.setup();
+        for (RoleResource r : ALL_ROLES)
+            roleManager.createRole(AuthenticatedUser.ANONYMOUS_USER, r, new RoleOptions());
+
+        CassandraRoleManager crm = new CassandraRoleManager();
+
+        assertEquals(CassandraRoleManager.hasExistingRoles(), true);
+        assertEquals(crm.isClusterReady(), false);
```
Comment (maedhroz) [inline test/unit/org/apache/cassandra/auth/CassandraRoleManagerTest.java]: nit: could use `assertTrue()` and `assertFalse()` respectively (although perhaps that's less important than throwing an explicit failure message in there)

File: test/unit/org/apache/cassandra/auth/PasswordAuthenticatorTest.java [hunk 18]
```diff
@@ -19,36 +19,58 @@
...
-import static org.mindrot.jbcrypt.BCrypt.hashpw;
 import static org.mindrot.jbcrypt.BCrypt.gensalt;
+import static org.mindrot.jbcrypt.BCrypt.hashpw;
+
+import static org.apache.cassandra.auth.CassandraRoleManager.GENSALT_LOG2_ROUNDS_PROPERTY;
 
 public class PasswordAuthenticatorTest extends CQLTester
 {
-
     private static PasswordAuthenticator authenticator = new PasswordAuthenticator();
```
Comment (maedhroz) [inline test/unit/org/apache/cassandra/auth/PasswordAuthenticatorTest.java]: nit: could be `final`

File: test/unit/org/apache/cassandra/auth/PasswordAuthenticatorTest.java [hunk 19]
```diff
@@ -161,18 +183,30 @@ private void testDecodeIllegalUserAndPwd(String username, String password)
         negotiator.getAuthenticatedUser();
     }
 
-    @BeforeClass
-    public static void setUp()
+    @Test
+    public void warmCacheLoadsAllEntriesFromTables() throws Exception
```
Comment (maedhroz) [inline test/unit/org/apache/cassandra/auth/PasswordAuthenticatorTest.java]: nit: could remove `throws Exception`

File: test/unit/org/apache/cassandra/auth/PasswordAuthenticatorTest.java [hunk 20]
```diff
@@ -161,18 +183,30 @@ private void testDecodeIllegalUserAndPwd(String username, String password)
...
+        Map<String, String> cacheEntries = authenticator.bulkLoader().get();
+
+        assertEquals(ALL_ROLES.length, cacheEntries.size());
+        cacheEntries.forEach((username, hash) -> assertTrue(BCrypt.checkpw("hash_for_" + username, hash)));
     }
 
-    @AfterClass
-    public static void tearDown()
+    @Test
+    public void warmCacheWithEmptyTable() throws Exception
```
Comment (maedhroz) [inline test/unit/org/apache/cassandra/auth/PasswordAuthenticatorTest.java]: nit: could remove `throws Exception`

File: src/java/org/apache/cassandra/service/StorageService.java (hunk 6 above)
Comment (jmckenzie-dev) [reply src/java/org/apache/cassandra/service/StorageService.java]: Bah! Something in my IDE is clearly misbehaving. Sorry for these stragglers; I'll get to the bottom of it.

File: src/java/org/apache/cassandra/auth/AuthCacheService.java (hunk 1 above)
Comment (jmckenzie-dev) [reply src/java/org/apache/cassandra/auth/AuthCacheService.java]: :+1:

File: src/java/org/apache/cassandra/config/Config.java (hunk 4 above)
Comment (jmckenzie-dev) [reply src/java/org/apache/cassandra/config/Config.java]: Hm. Nope, should be optional. Changed to false and added to cassandra.yaml

File: test/unit/org/apache/cassandra/tools/nodetool/InvalidateNetworkPermissionsCacheTest.java (hunk 9 above)
Comment (jmckenzie-dev) [reply test/unit/org/apache/cassandra/tools/nodetool/InvalidateNetworkPermissionsCacheTest.java]: Not at all. Removed.

File: test/unit/org/apache/cassandra/auth/AuthCacheTest.java (hunk 13 above)
Comment (jmckenzie-dev) [reply test/unit/org/apache/cassandra/auth/AuthCacheTest.java]: Forklifted to a class level final int constant. :+1:

File: test/unit/org/apache/cassandra/auth/CassandraRoleManagerTest.java (hunk 16 above)
Comment (jmckenzie-dev) [reply test/unit/org/apache/cassandra/auth/CassandraRoleManagerTest.java]: ! Missed one. Got a few others. :+1:

File: test/unit/org/apache/cassandra/auth/CassandraRoleManagerTest.java (hunk 17 above)
Comment (jmckenzie-dev) [reply test/unit/org/apache/cassandra/auth/CassandraRoleManagerTest.java]: Fleshed out a bit and swapped to the less tautological options.

File: test/unit/org/apache/cassandra/auth/PasswordAuthenticatorTest.java (hunk 18 above)
Comment (jmckenzie-dev) [reply test/unit/org/apache/cassandra/auth/PasswordAuthenticatorTest.java]: Interesting. We had a couple PasswordAuthenticator creations in the tests that shadowed the class static variable (ew), some vestigial throws Exception, a whitespace... and this. Tidied that file up.

File: src/java/org/apache/cassandra/auth/AuthCacheService.java (hunk 3 above)
Comment (jmckenzie-dev) [reply src/java/org/apache/cassandra/auth/AuthCacheService.java:38]: I think we should protect against dupe adding from all 4 sources as well as synchronize on the init and reg call. I'll whip that up and get another commit for that; forgot to come back to this.

Comment (jmckenzie-dev) [reply src/java/org/apache/cassandra/auth/AuthCacheService.java:38]: Pushed a commit that hopefully addresses this. Running CI now.

File: src/java/org/apache/cassandra/auth/CassandraRoleManager.java [hunk 21]
```diff
@@ -135,6 +135,12 @@ static int getGensaltLogRounds()
     // Will be set to true when all nodes in the cluster are on a version which supports roles (i.e. 2.2+)
     private volatile boolean isClusterReady = false;
```
Comment (maedhroz) [inline src/java/org/apache/cassandra/auth/CassandraRoleManager.java]: Can we just remove `isClusterReady` altogether? We won't allow upgrading directly from 2.1 -> 4.1 right?

File: src/java/org/apache/cassandra/auth/AuthCacheService.java (hunk 3 above)
Comment (maedhroz) [reply src/java/org/apache/cassandra/auth/AuthCacheService.java:38]: `AuthCacheService` is at least thread-safe now, regardless of whether or not that's strictly necessary. The only thing that's still a bit odd to me is that the `nodetool` commands (unless I'm misreading something) can register caches that just sort of sit in `caches` forever, being neither removed nor warmed.

Comment (jmckenzie-dev) [reply src/java/org/apache/cassandra/auth/AuthCacheService.java:38]: That... does not seem like something we want. Let me take a look.

Comment (jmckenzie-dev) [reply src/java/org/apache/cassandra/auth/AuthCacheService.java:38]: Ok. Taking a page out of the guard around doAuthSetup to codify the intended usage of the method.

**Review by maedhroz — APPROVED** [review]
